### PR TITLE
Remove DSA support

### LINKED ISSUE #2454: Removing DSA support
### Problem Description

There is a lot of generic DSA support throughout the library, but my fast grep shows that the DSA keys are *somehow* handled only in the GPK, epass2003 and oberthur drivers:
```
src/pkcs15init/pkcs15-epass2003.c:              case SC_PKCS15_TYPE_PUBKEY_DSA:
```
 * just pubkey template/names -- probably not working
```
src/pkcs15init/pkcs15-gpk.c:    case SC_PKCS15_TYPE_PRKEY_DSA:
src/pkcs15init/pkcs15-gpk.c:            algo = SC_ALGORITHM_DSA; break;
```
 * this looks almost like complete support
```
src/pkcs15init/pkcs15-oberthur.c:               case SC_PKCS15_TYPE_PUBKEY_DSA:
```
 * just public keys template/names -- probably not working
```
src/libopensc/cardctl.h:        SC_CARDCTL_OBERTHUR_KEY_DSA_PUBLIC,
src/libopensc/cardctl.h:        SC_CARDCTL_OBERTHUR_KEY_DSA_PRIVATE,
```
 * unused defines
```
src/libopensc/card-dnie.c:              case SC_ALGORITHM_DSA:
```
 * just used to say that it is not supported

I think this might work only in the GPK driver to some extent, but that one is unmaintained anyway since 207 or so and planned to be [deactivated](https://github.com/OpenSC/OpenSC/wiki/Removal-of-Old-Card-Drivers#card-driver-overview).

### Proposed Resolution

I am all for removing the support for the DSA in libopensc. It is for no use today and is just maintenance burden. It is not even supported by its creator, NIST in latest FIPS 140-3 release after 2023 so. Before submitting a PRs, I just wanted to check if there are some strong voices against doing so.

I am not about slashing all the occurrences of the DSA (I am fine if it can be detected in pkcs11-tool if needed or similar), but the internal handling this key type in libopensc is not useful.

### PR DESCRIPTION
**Pull request by xhanulik**
Fixes  #2454

PR removes support for DSA, mostly generic code, `pkcs15-init`, `pkcs15-tool` and the support in GPK and epass2003 drivers.

##### Checklist
- [x] Documentation is added or updated
- [ ] PKCS#11 module is tested
- [ ] Windows minidriver is tested
- [ ] macOS tokend is tested

### CODE REVIEW

File: src/pkcs15init/pkcs15-entersafe.c [hunk 1]
```diff
@@ -355,10 +355,7 @@ static int entersafe_store_key(sc_profile_t *profile, sc_pkcs15_card_t *p15card,
 	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
 
 	if ( key->algorithm != SC_ALGORITHM_RSA )
-	{
-		 /* ignore DSA keys */
```
Comment (Jakuje) [inline src/pkcs15init/pkcs15-entersafe.c:359]: I would keep this comment, maybe just rewrite it to "ignore non-RSA keys"

File: src/tools/pkcs11-tool.c [hunk 2]
```diff
@@ -6035,7 +6035,6 @@ static EVP_PKEY *get_public_key(CK_SESSION_HANDLE session, CK_OBJECT_HANDLE priv
 			OSSL_PARAM_free(params);
 #endif
 			return pkey;
-		case CKK_DSA:
```
Comment (Jakuje) [inline src/tools/pkcs11-tool.c:6038]: Maybe we could keep this one if somebody would need to use the pkcs11-tool with some other module that supports DSA.

**Review by Jakuje — APPROVED** [review]

**Comment by Jakuje** [comment]
@OpenSC/maintainers any more concerns with this change? Is there anyone who has some card with DSA that is working?

**Comment by Jakuje** [comment]
Any strong opinions if we should get this in now or after the next release?

So far I did not hear any reasoning why not to remove DSA support, but remaining question is when. I think this PR is in a good shape to be merged, but I do not insist on doing it right away somebody feels an urge to push this through some announcements.

But I hope nobody in 2022 is using 1024 bits DSA keys. I still think the current implementation never worked in PKCS#11 as the DSA algorithm was never registred in `_sc_card_add_algorithm` in any card driver and there was no support in the pkcs11-tool either.

File: src/libopensc/pkcs15.h [hunk 3]
```diff
@@ -448,19 +428,17 @@ typedef struct sc_pkcs15_skey_info sc_pkcs15_skey_info_t;
...
+#define SC_PKCS15_TYPE_PRKEY_XEDDSA		0x105
 
 #define SC_PKCS15_TYPE_PUBKEY			0x200
 #define SC_PKCS15_TYPE_PUBKEY_RSA		0x201
-#define SC_PKCS15_TYPE_PUBKEY_DSA		0x202
-#define SC_PKCS15_TYPE_PUBKEY_GOSTR3410		0x203
-#define SC_PKCS15_TYPE_PUBKEY_EC		0x204
-#define SC_PKCS15_TYPE_PUBKEY_EDDSA		0x205
-#define SC_PKCS15_TYPE_PUBKEY_XEDDSA		0x206
+#define SC_PKCS15_TYPE_PUBKEY_GOSTR3410		0x202
```
[the remark this answers is not in the document]
Comment (xhanulik) [reply src/libopensc/pkcs15.h]: As @Jakuje mentioned above, it should probably stay as it was originally, it's fixed already.

**Review by frankmorgner — APPROVED** [review]
I think we're good to include this in the next release, too.

File: src/libopensc/pkcs15.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jakuje) [reply src/libopensc/pkcs15.h]: With this change reverted, we need to adjust also the pkcs15-tool change, which is using these constants for key type labels.

[the remark this answers is not in the document]
Comment (xhanulik) [reply src/libopensc/pkcs15.h]: Thanks for pointing out, fixed.

**Pull request merged**